docs/TPPDialect.md

```diff
@@ -0,0 +1,355 @@
+
+After bufferization, the dialect also supports `memref` types, where further grouping, transforms and annotation happen, in preparaiton for optimal `xsmm` lowering.
+
+## Optimizations
+
+Due to cache locality, instruction set availability and number of registers, there are various costs and benefits of picking particular combinations, which aren't immediately clear to the programmer.
+
+Our compiler needs to select the most appropriate combination that can be lowered to existing and optiomal `libxsmm` calls with the right blockng and tiling factors.
+
+TPP operations can carry attributes, their operands can be chained and grouped, forming trees and graphs, which can be mathed in different ways.
+
+Once lowered to `xsmm` dialect, the operations are selected and any further lowering won't change the TPP semantics, only add `xsmm` boilerplate and infrastruture to execute the chosen strategies.
+
+# Basic Rules
+
+## Syntax
+
+All `tensor` operations follow the SSA syntax, operands and return values:
+```mlir
+  // Always out-of-place
+  %C = tpp.add ins(%arg0, %arg1) : (tensor<>, tensor<>) -> tensor<>
+  // %C is here treated as initialization, accumulation can still happen on same buffer
+  %M = tpp.matmul ins(%A, %B, %C) : (tensor<>, tensor<>, tensor<>) -> tensor<>
+```
+
+All `memref` operations follow the `linalg` syntax of `ins` and `outs`.
+The `outs` values are an indication of where the output is written to.
+These are generated by the bufferization process, from the `tensor` level operations.
+```mlir
+  // Out-of-place ops output on new allocations
+  %buf = memref.alloc : memref<>
+  tpp.add ins(%arg0, %arg1) outs(%buf) : (memref<>, memref<>)
+  // In-of-place ops output on input operands
+  tpp.matmul ins(%A, %B, %C) outs(%C) : (memref<>, memref<>, memref<>)
+```
+
+### Unary
+
+On `tensors`, unary ops have one argument and one return value.
+
+On `memrefs`, unary operations have only `outs` for non-transforms (neg, inv, exp, etc.).
+For transforms (`broadcast`, `reduce`, `transpose`, `pack`, etc), there is one `ins` argument and one `outs` argument.
+The sizes must be the same for element-wise / arith, but will often be different for transforms (except `copy` and `transpose`).
+
+There is no implicit transform semantics.
+When lowering a broadcast (ex. a `linalg.generic` with lower dimension argument and apporpriate affine map), we map to `tpp.broadcast` + the operation.
+
+### Binary
+
+On `tensors`, binary ops have two arguments and one return value.
+
+On `memrefs`, binary operations have both `ins` and `outs`, with always two `ins` and one `outs` arguments.
+On most binary ops, the `outs` can alias `ins` (ex. `a=a+b`), and `ins` can alias themselves (ex. `a=a+a`).
+
+The exception is `matmul`, where the output cannot alias the inputs.
+A binary `matmul` is equivalent to `C = A x B`, with `C` as a new `memref` allocation.
```

Comment:
We probably need to also add zero initialization to new buffer allocation or reuse.

Reply:
Which transforms this into a ternary op, reinforcing my point below. :)

Reply:
Agreed ^